sdk/provisioning/Azure.Provisioning/src/ModuleConstruct.cs

```diff
@@ -52,12 +61,209 @@ private static ConstructScope ResourceToConstructScope(Resource resource)
+                // regular resource in anonymous scope
+                _ => ConstructScope.ResourceGroup,
             };
         }
+
+        public BinaryData SerializeModuleReference()
+        {
+            using var stream = new MemoryStream();
+            stream.WriteLine($"module {Name} './resources/{Name}/{Name}.bicep' = {{");
+            stream.WriteLine($"  name: '{Name}'");
+            stream.WriteLine($"  scope: {GetScopeName()}");
+
+            var parametersToWrite = new HashSet<Parameter>();
+            var outputs = new HashSet<Output>(GetOutputs());
+            foreach (var p in GetParameters(false))
+            {
+                if (!ShouldExposeParameter(p, outputs))
+                {
+                    continue;
+                }
+                parametersToWrite.Add(p);
+            }
+            if (parametersToWrite.Count > 0)
+            {
+                stream.WriteLine($"  params: {{");
+                foreach (var parameter in parametersToWrite)
+                {
+                    stream.WriteLine($"    {parameter.Name}: {parameter.GetParameterString(Scope!)}");
+                }
+                stream.WriteLine($"  }}");
+            }
+            stream.WriteLine($"}}");
+
+            return new BinaryData(stream.GetBuffer().AsMemory(0, (int)stream.Position));
+        }
+
+        public BinaryData SerializeModule()
+        {
+            using var stream = new MemoryStream();
+            var options = new ModelReaderWriterOptions("bicep");
+
+            WriteScopeLine(stream);
+
+            WriteParameters(stream);
+
+            WriteExistingResources(stream);
+
+            foreach (var resource in GetResources(false))
+            {
+                if (resource is Tenant)
+                {
+                    continue;
+                }
+                stream.WriteLine();
+                WriteLines(0, ModelReaderWriter.Write(resource, options), stream, resource);
+            }
+
+            foreach (var construct in GetConstructs(false).Select(c => (ModuleConstruct)c))
+            {
+                stream.WriteLine();
+                stream.Write(construct.SerializeModuleReference().ToArray());
+            }
+
+            WriteOutputs(stream);
+
+            return new BinaryData(stream.GetBuffer().AsMemory(0, (int)stream.Position));
+        }
+
+        private void WriteExistingResources(MemoryStream stream)
+        {
+            foreach (var resource in GetExistingResources(false))
+            {
+                stream.WriteLine();
+                stream.WriteLine($"resource {resource.Name} '{resource.Id.ResourceType}@{resource.Version}' existing = {{");
+                stream.WriteLine($"  name: '{resource.Name}'");
+                stream.WriteLine($"}}");
+            }
+        }
+
+        private void WriteScopeLine(MemoryStream stream)
+        {
+            if (ConstructScope != ConstructScope.ResourceGroup || IsRoot)
```

Comment:
The serialization functionality was moved into ModuleConstruct as there is no use case for serializing a non-module construct. Also this allows to access ModuleConstruct properties like IsRoot.